code/modules/projectiles/guns/energy.dm

```diff
@@ -411,3 +411,129 @@
 
 /obj/item/weapon/gun/energy/lasgun/M43/practice/unique_action(mob/user)
 	return
+
+/obj/item/weapon/gun/energy/lasgun/lasrifle
+	name = "\improper TX-78 Sunfury Lasgun MK1"
+	desc = "An multifunctional laser based rifle with an integrated mode selector. Ideal for any situation. Uses power cells instead of ballistic magazines.."
+	force = 20 //Large and hefty! Includes stock bonus.
+	icon = 'icons/Marine/lasrifle.dmi'
+	icon_state = "tx73"
+	item_state = "tx73"
+	max_shots = 50 //codex stuff
+	load_method = CELL //codex stuff
+	ammo = /datum/ammo/energy/lasgun/M43
+	ammo_diff = null
+	cell_type = null
+	charge_cost = M43_STANDARD_AMMO_COST
+	attachable_allowed = list(
+						/obj/item/attachable/bayonet,
+						/obj/item/attachable/reddot,
+						/obj/item/attachable/verticalgrip,
+						/obj/item/attachable/angledgrip,
+						/obj/item/attachable/lasersight,
+						/obj/item/attachable/gyro,
+						/obj/item/attachable/flashlight,
+						/obj/item/attachable/bipod,
+						/obj/item/attachable/magnetic_harness,
+						/obj/item/attachable/attached_gun/grenade,
+						/obj/item/attachable/scope,
+						/obj/item/attachable/attached_gun/flamer,
+						/obj/item/attachable/attached_gun/shotgun,
+						/obj/item/attachable/scope/mini)
+	flags_gun_features = GUN_AUTO_EJECTOR|GUN_CAN_POINTBLANK|GUN_AMMO_COUNTER|GUN_ENERGY|GUN_AMMO_COUNTER
+	attachable_offset = list("muzzle_x" = 32, "muzzle_y" = 18,"rail_x" = 12, "rail_y" = 23, "under_x" = 23, "under_y" = 15, "stock_x" = 22, "stock_y" = 12)
+
+	accuracy_mult_unwielded = 0.5 //Heavy and unwieldy; you don't one hand this.
+	scatter_unwielded = 100 //Heavy and unwieldy; you don't one hand this.
+	damage_falloff_mult = 0.25
+	fire_delay = 3
+	var/current_mode = 1
```

Comment:
Use human-readable defines for the modes.

Reply:
Gonna argue against that. We have already gone though a patch where every gun value was defined. And we know that got changed quite a while back.

On the point of `current_mode`, there are no switch statements, it's purely used as a counter to keep track of current position.

Reply:
Does it track different fire modes or something purely numeric?

Reply:
purely numeric. it tracks the position in the list of firemodes.

As nice as STANDARD_FIREMODE 1 and AUTO_FIREMODE 2 would be, we'd only ever actually have the STANDARD_FIREMODE in code.

Reply:
rename it to mode_index or something